core/src/main/java/brooklyn/event/basic/MapConfigKey.java

```diff
@@ -125,6 +130,20 @@ protected Object applyEntryValueToMap(Entry value, Map target) {
         } else if (k instanceof String) {
             k = subKey((String)k);
         } else {
+            // supplier or other unexpected value
+            if (k instanceof Supplier) {
+                // TODO not thread-safe
+                Object mapAtRoot = target.get(this);
+                if (mapAtRoot==null) {
+                    mapAtRoot = new LinkedHashMap();
+                    target.put(this, mapAtRoot);
+                }
+                if (mapAtRoot instanceof Map) {
+                    synchronized (mapAtRoot) {
```

Comment:
often the map will be what we `put` just above -- ie an LHMap, and sometimes it will come from another source, e.g. yaml parse (also often an LHMap).

in most cases synching on the map is the standard way to synchronize.  it's not foolproof of course.  but feels likely to cause a problem than a CME on an LHMap.

will add a comment to that effect.